Add options to kick & ban users from community groups

mrdaak opened the pull request:
Fixes #13872 

### Summary

Add kick and ban options for community admins on long press over: text, image, audio messages.
Please, see issue for more details.

#### Platforms

- Android
- iOS

#### Areas that maybe impacted
Community chat, group chat

### Steps to test
As a community admin, long press over a message in a community chat, and select `Kick member` or `Ban member`.

status: ready

mrdaak committed:
Add options to kick & ban users from community groups

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 8d9f0986 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-14276/1/) | 2022-10-30 19:09:15 | ~2 min | `tests` | [:package:`log`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Mobile-221030-190626-8d9f09-pr14276-tests.log) |

du64 commented:
It shouldn't be possible to ban the community owner.
I think we should also seperate community actions from deleting messages, to prevent accidental bans

![Screenshot_2022-10-30-15-33-22-08_65ad86f86fd2321a85272a99323da10f](https://user-images.githubusercontent.com/65834069/198900362-ac1b53cd-f789-4798-a05c-f3883bafa45b.jpg)

ibrkhalil commented on src/status_im/ui2/screens/chat/messages/view.cljs:
```diff
@@ -130,6 +132,8 @@
      :public?             public?
      :community?          (not (nil? community-id))
+     :community-id        community-id
+     :can-manage-users?   (:can-manage-users? community)
```
Destructure please

mrdaak replied:
Applied. Thanks!

ibrkhalil commented on src/status_im/ui2/screens/chat/messages/message.cljs:
```diff
@@ -617,14 +659,25 @@
+                                                    :icon     :main-icons2/delete
+                                                    :id       :delete})
+                                                 (when (and community? can-manage-users?)
+                                                   [{:type     :danger
```
Why are we returning a vector .. Shouldn't we be concatenating into the original one ?

mrdaak replied:
Yes, I missed it out. It's all concatenated now into the original one. 
Looks good?

ibrkhalil replied:
Thanks, Yup!

ilmotta commented on src/status_im/ui2/screens/chat/messages/message.cljs:
```diff
@@ -554,40 +585,51 @@
-                                     :id       :delete}]))))]
+                                     :icon     :main-icons2/delete
+                                     :id       :delete}])
+                                 (when (and community? can-manage-users?)
```
This logical check is repeated a couple times throughout this file, but it has an important meaning. How about we extract it into a function named `allow-dangerous-community-actions?` (or any other meaningful name). As @du64 mentioned in https://github.com/status-im/status-mobile/pull/14276#issuecomment-1296345570, this function could also take care of checking community owners cannot be kicked/banned.

mrdaak replied:
Hey Icaro!

I've followed your suggestion and extracted it into a separate function. Thanks for the feedback!
I am only missing a check if message sender is an admin. 

1) `get-render-data` receives `admins` but it is always empty, so I assume this is not intended for *community chats*.  
2) Looking into `(:members community)`, I assume:
- *admin* has `{:roles [1]}` 
- and *regular user* just `{}`. 

Is there a dedicated logic for this? Or I can make a subscriber If my assumptions are correct.

ilmotta replied:
Hey Miloš!

Thanks for taking the effort to refactor. Unfortunately your questions are too specific for me, because I haven't been working in the communities epic. I'm looping in other devs to see if they can help.

@cammellos, @jo-mut, @J-Son89, perhaps one of you can help @mrdaak reliably detect if the sender is a community group owner/admin, so that we can avoid banning/kicking them.

cammellos replied:
I'd say not to worry about kicking banning the owner, he has full control of the community, so even if it's kicked/banned, he's able to join it back. Also it would not impact the management as that's done through the community keypair, so it should be fine, but we can look more into it in a separate PR, I think for this PR, trusting `can-manage-users?` is enough

mrdaak replied:
@cammellos thanks for the feedback! Given that `can-manage-users?` is enough, I'd say PR is ready from my side.
If there is more space for improvement, please let me know. :)

ilmotta commented on src/status_im/ui2/screens/chat/messages/message.cljs:
```diff
@@ -394,7 +395,15 @@
+                                              community?
+                                              community-id
+                                              can-manage-users?
+                                              from] :as message} content]
```
[Code style] The Clojure Style Guide recommends breaking arguments over multiple lines for readability purposes. So something like this for the arguments' vector:

```clojure
[on-long-press
 {:keys [pinned
         message-pin-enabled
         outgoing
         edit-enabled
         show-input?
         community?
         community-id
         can-manage-users?
         from]
  :as   message}
 content]
```

ilmotta commented on src/status_im/ui2/screens/chat/messages/view.cljs:
```diff
@@ -118,10 +120,10 @@
-        community (<sub [:communities/community community-id])
-        group-admin? (get admins current-public-key)
-        community-admin? (when community (community :admin))
+  (let [current-public-key  (<sub [:multiaccount/public-key])
```
Thanks for aligning the let bindings :+1:, but could you also break the long destructuring of map keys into multiple lines? It's almost wider than my screen ;)

mrdaak replied:
Done. 👍
Talking of the style guide, I've noticed the `:require` namespaces are not sorted. It looks like a nice fix, and I wonder If you have an Emacs shortcut for that. :)

ilmotta replied:
Good point. Nowadays the linter doesn't force us to sort namespaces. This is going to change if/when we integrate zprint in the codebase. For now, you can see there's a lot of work in this draft PR https://github.com/status-im/status-mobile/pull/14128

> I wonder If you have an Emacs shortcut for that

Emacs by itself doesn't support anything like that, but if you have installed and configured the `lsp` package https://github.com/emacs-lsp/lsp-mode, then you can call the command `lsp-clojure-clean-ns`, or define a keybinding for that (that's what I do).

mrdaak replied:
Thanks for the pointers! I'll dig more into the `lsp-mode`.

mrdaak committed:
Address review feedback

Separate kick & ban actions in admin section
Replace placeholder icons with dedicated ones: arrow-left & block
Follow the Clojure Style Guide (TODO sort namespaces)

mrdaak committed:
Format

mrdaak committed:
Don't kick yourself out

churik commented:
@mrdaak thank you for PR!
Please, rebase it to the current develop, make sure that all comments are addressed and `make tests` passed.

ibrkhalil reviewed (approved):
lgtm

mrdaak committed:
...

mrdaak commented:
> @mrdaak thank you for PR! Please, rebase it to the current develop, make sure that all comments are addressed and `make test` passed.

@churik PR is rebased and all comments addressed.
```make test``` returns 0 failures, 0 errors
<img width="287" alt="image" src="https://user-images.githubusercontent.com/22012501/204370122-a5e05404-a307-43fd-9eb6-ed632e657541.png">

cammellos commented:
> > @mrdaak thank you for PR! Please, rebase it to the current develop, make sure that all comments are addressed and `make test` passed.
> 
> @churik PR is rebased and all comments addressed. `make test` returns 0 failures, 0 errors <img alt="image" width="287" src="https://user-images.githubusercontent.com/22012501/204370122-a5e05404-a307-43fd-9eb6-ed632e657541.png">

Hi, sorry, it's `make lint` that fails:

```
[2022-11-28T19:26:14.741Z] src/status_im/ui2/screens/chat/messages/message.cljs:421:70: warning: unused binding message


```
Thanks for your contribution!

mrdaak committed:
Fix a linter error

mrdaak commented:
> Hi, sorry, it's `make lint` that fails:

Fixed. All green 👍 

> Thanks for your contribution!

Thanks for the opportunity! @cammellos

status-im-auto commented:
## 100% of end-end tests have passed
```
Total executed tests: 7
Failed tests: 0
Passed tests: 7
```
<h3>Passed tests (7)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestDeeplinkOneDeviceNewUI:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3059383">test_public_chat_open_using_deep_link</a>, id: 702776 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/fcab1c87d73643218f46b6c9df81938f?auth=1ac47bfd47e6b214ea1a1f48452147f8#79">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3059381">test_deep_link_with_invalid_user_public_key_own_profile_key</a>, id: 702774 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/fcab1c87d73643218f46b6c9df81938f?auth=1ac47bfd47e6b214ea1a1f48452147f8#40">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3059382">test_deep_link_open_user_profile</a>, id: 702775 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/fcab1c87d73643218f46b6c9df81938f?auth=1ac47bfd47e6b214ea1a1f48452147f8#97">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3059379">test_1_1_chat_text_message_delete_push_disappear</a>, id: 702733 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/e5d614fd14d949d0b5bc9540755505e1?auth=8489d7b6ca57105b04320947a4a8f02d#255">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/3578c8a87f314623881819088c28bab0?auth=7760fc1e022422f34ef69e3199827dd7#171">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3059380">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/e5d614fd14d949d0b5bc9540755505e1?auth=8489d7b6ca57105b04320947a4a8f02d#176">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/3578c8a87f314623881819088c28bab0?auth=7760fc1e022422f34ef69e3199827dd7#104">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3059377">test_1_1_chat_message_reaction</a>, id: 702730 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/e5d614fd14d949d0b5bc9540755505e1?auth=8489d7b6ca57105b04320947a4a8f02d#66">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/3578c8a87f314623881819088c28bab0?auth=7760fc1e022422f34ef69e3199827dd7#53">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>4. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3059378">test_1_1_chat_pin_messages</a>, id: 702731 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/e5d614fd14d949d0b5bc9540755505e1?auth=8489d7b6ca57105b04320947a4a8f02d#86">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/3578c8a87f314623881819088c28bab0?auth=7760fc1e022422f34ef69e3199827dd7#74">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details>

churik commented:
I can't verify Pr, as for some reason community invite is not visible there.
The issue is not reproducible on nighlty 29/11/22

**Steps:**
- create a community ("Require approval" type)
- invite there your contact
**Expected result:**
- can see the invite in chat and join the community
**Actual result:**
https://user-images.githubusercontent.com/4557972/204542708-0a4a2612-6deb-4d53-9560-4fff31623c90.mp4

VladimrLitvinenko commented:
> I can't verify Pr, as for some reason community invite is not visible there. The issue is not reproducible on nighlty 29/11/22
> 
> **Steps:**
> 
> * create a community ("Require approval" type)
> * invite there your contact
>   **Expected result:**
> * can see the invite in chat and join the community
>   **Actual result:**
>   https://user-images.githubusercontent.com/4557972/204542708-0a4a2612-6deb-4d53-9560-4fff31623c90.mp4

Hi @mrdaak were you able to reproduce this issue?

mrdaak committed:
..

mrdaak commented:
> Hi @mrdaak were you able to reproduce this issue?

I did, and it also happens on the `develop` branch for me, but I was unable to figure out why, and would appreciate any help to debug it.

Branch is rebased, and `make lint` and `make test` pass on my side, not sure why `prs/tests` fails on CI.
Keeping up with the changes is hard, I'd appreciate assistance for this PR to get closed. Thanks

du64 commented:
I meant to delete a users message and accidentally kicked them. is there a UI or other fix for this?

flexsurfer commented:
hey @mrdaak thanks for your contribution, are you going to continue work on this ?
